yw_website/yw_website/apps/website/serializers.py

```diff
@@ -157,25 +283,180 @@ def __create_tags(self, w, validated_data):
+            validated_data.get("programBlock"), many=True, context={"run": r}
+        )
+        for program_block in program_blocks.data:
+            try:
+                in_block = ProgramBlock.objects.get(
+                    programblock_id=program_block.get("inProgramBlock"), run=r
+                )
+            except ProgramBlock.DoesNotExist:
+                in_block = None
+
+            pb = ProgramBlock(
+                programblock_id=program_block.get("programBlockId"),
+                name=program_block.get("name"),
+                qualified_name=program_block.get("qualifiedName"),
+                in_program_block=in_block,
+                run=r,
+            )
+            pb.save()
+
+    def _create_data(self, r, validated_data):
+        datas = DataSerializer(
+            validated_data.get("data"), many=True, context={"run": r}
+        )
+        for data in datas.data:
+            try:
+                in_block = ProgramBlock.objects.get(
+                    programblock_id=data.get("inProgramBlock"), run=r
+                )
+            except ProgramBlock.DoesNotExist:
+                in_block = None
+
+            d = Data(
+                data_id=data.get("dataId"),
+                name=data.get("name"),
+                qualified_name=data.get("qualifiedName"),
+                in_program_block=in_block,
+                run=r,
+            )
+            d.save()
+
+    def _create_ports(self, r, validated_data):
+        ports = PortSerializer(
+            validated_data.get("port"), many=True, context={"run": r}
+        )
+        for port in ports.data:
+            try:
+                in_block = ProgramBlock.objects.get(
+                    programblock_id=port.get("inProgramBlock"), run=r
+                )
+            except ProgramBlock.DoesNotExist:
+                in_block = None
+
+            try:
+                data = Data.objects.get(data_id=port.get("data"), run=r)
+            except Data.DoesNotExist:
+                data = None
+
+            p = Port(
+                port_id=port.get("portId"),
+                in_program_block=in_block,
+                data=data,
+                name=port.get("name"),
+                qualified_name=port.get("qualifiedName"),
+                alias=port.get("alias"),
+                uri_template=port.get("uriTemplate"),
+                is_inport=port.get("inPort"),
+                is_outport=port.get("outPort"),
+                run=r,
+            )
+            p.save()
+
+    def _create_channels(self, r, validated_data):
+        channels = ChannelSerializer(validated_data.get("channel"), many=True)
+        for channel in channels.data:
+            try:
+                out_port = Port.objects.get(port_id=channel.get("outPort"), run=r)
+                in_port = Port.objects.get(port_id=channel.get("inPort"), run=r)
+                data = Data.objects.get(data_id=channel.get("data"), run=r)
```

Comment:
a channel should also always be associated with an inport, outport, and data. otherwise we throw a 500